x/bsonx/bsoncore/bsonbuilder.go

```diff
@@ -0,0 +1,362 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bsoncore
+
+import (
+	"strconv"
+
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+// DocumentBuilder builds a bson document
+type DocumentBuilder struct {
+	doc     []byte
+	indexes []int32
+}
+
+// ArrayBuilder builds a bson array
+type ArrayBuilder struct {
+	arr     []byte
+	indexes []int32
+	key     []int
+}
+
+// startDocument reserves the document's length and set the index to where the length begins
+func (db *DocumentBuilder) startDocument() *DocumentBuilder {
+	var index int32
+	index, db.doc = AppendDocumentStart(db.doc)
+	db.indexes = append(db.indexes, index)
+	return db
+}
+
+// startArray reserves the array's length and sets the index to where the length begins
+func (a *ArrayBuilder) startArray() *ArrayBuilder {
+	var index int32
+	index, a.arr = AppendArrayStart(a.arr)
+	a.indexes = append(a.indexes, index)
+	a.key = append(a.key, 0)
+	return a
+}
+
+// NewDocumentBuilder creates a new DocumentBuilder
+func NewDocumentBuilder() *DocumentBuilder {
+	return (&DocumentBuilder{}).startDocument()
+}
+
+// NewArrayBuilder creates a new ArrayBuilder
+func NewArrayBuilder() *ArrayBuilder {
+	return (&ArrayBuilder{}).startArray()
+}
+
+// Build updates the length of the document and index to the beginning of the documents length
+// bytes, then returns the document (bson bytes)
+func (db *DocumentBuilder) Build() Document {
+	last := len(db.indexes) - 1
+	db.doc = append(db.doc, 0x00)
+	db.doc = UpdateLength(db.doc, db.indexes[last], int32(len(db.doc[db.indexes[last]:])))
+	db.indexes = db.indexes[:last]
+	return db.doc
+}
+
+// Build updates the length of the array and index to the beginning of the documents length
+// bytes, then returns the array (bson bytes)
+func (a *ArrayBuilder) Build() Array {
+	last := len(a.indexes) - 1
+	a.arr = append(a.arr, 0x00)
+	a.arr = UpdateLength(a.arr, a.indexes[last], int32(len(a.arr[a.indexes[last]:])))
+	a.indexes = a.indexes[:last]
+	a.key = a.key[:len(a.key)-1]
+	return a.arr
+}
+
+// AppendInt32 will append an int32 element using key and i32 to DocumentBuilder.doc
+func (db *DocumentBuilder) AppendInt32(key string, i32 int32) *DocumentBuilder {
+	db.doc = AppendInt32Element(db.doc, key, i32)
+	return db
+}
+
+// AppendInt32 will append i32 to ArrayBuilder.arr
+func (a *ArrayBuilder) AppendInt32(i32 int32) *ArrayBuilder {
+	last := len(a.key) - 1
```

Comment:
This code is duplicated in all of the `ArrayBuilder.Append*` functions:

```
last := len(a.key)-1
// use strconv.Itoa(a.key[last]) as a key
a.key[last]++
```

Can you refactor a bit to avoid code duplcation?

Reply:
Just from a cursory reading, I don't think the `appendElement` approach is readable. Also, it adds a dependency on `reflect`, which we want to avoid because this should be a high-performance code path. I think it'd be more readable if we just refactored the bit that gets the key and increments it into a function:

```
func (a *arrayBuilder) incrementKey() string {
    idx := len(a.key) - 1
    key := strconv.Itoa(a.key[idx])
    a.key[idx]++
    return key
}

func (a *arrayBuilder) AppendInt32(i32 int32) *ArrayBuilder {
    a.arr = AppendDoubleElement(a.arr, a.incrementKey(), i32)
}
```

What do you think?

Reply:
I think this solution works better. Updated